pkg/resource/cluster/update_post_build.go

```diff
@@ -0,0 +1,123 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package cluster
+
+import (
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	svcsdk "github.com/aws/aws-sdk-go/service/memorydb"
+)
+
+func (rm *resourceManager) newMemoryDBClusterUploadPayload(
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) *svcsdk.UpdateClusterInput {
+	res := &svcsdk.UpdateClusterInput{}
+
+	if delta.DifferentAt("Spec.AclName") && desired.ko.Spec.ACLName != nil {
+		res.SetACLName(*desired.ko.Spec.ACLName)
+	}
+	if desired.ko.Spec.Name != nil {
+		res.SetClusterName(*desired.ko.Spec.Name)
+	}
+	if delta.DifferentAt("Spec.Description") && desired.ko.Spec.Description != nil {
+		res.SetDescription(*desired.ko.Spec.Description)
+	}
+	if delta.DifferentAt("Spec.MaintenanceWindow") && desired.ko.Spec.MaintenanceWindow != nil {
+		res.SetMaintenanceWindow(*desired.ko.Spec.MaintenanceWindow)
+	}
+	if delta.DifferentAt("Spec.ParameterGroupName") && desired.ko.Spec.ParameterGroupName != nil {
+		res.SetParameterGroupName(*desired.ko.Spec.ParameterGroupName)
+	}
+	if delta.DifferentAt("Spec.SecurityGroupIDs") && desired.ko.Spec.SecurityGroupIDs != nil {
+		f8 := []*string{}
+		for _, f8iter := range desired.ko.Spec.SecurityGroupIDs {
+			var f8elem string
+			f8elem = *f8iter
+			f8 = append(f8, &f8elem)
+		}
+		res.SetSecurityGroupIds(f8)
+	}
+	if delta.DifferentAt("Spec.SnapshotRetentionLimit") && desired.ko.Spec.SnapshotRetentionLimit != nil {
+		res.SetSnapshotRetentionLimit(*desired.ko.Spec.SnapshotRetentionLimit)
+	}
+	if delta.DifferentAt("Spec.SnapshotWindow") && desired.ko.Spec.SnapshotWindow != nil {
+		res.SetSnapshotWindow(*desired.ko.Spec.SnapshotWindow)
+	}
+	if delta.DifferentAt("Spec.SnsTopicARN") && desired.ko.Spec.SnsTopicARN != nil {
+		res.SetSnsTopicArn(*desired.ko.Spec.SnsTopicARN)
+	}
+	if delta.DifferentAt("Spec.SnsTopicStatus") && desired.ko.Status.SnsTopicStatus != nil {
+		res.SetSnsTopicStatus(*desired.ko.Status.SnsTopicStatus)
+	}
+
+	if delta.DifferentAt("Spec.EngineVersion") && desired.ko.Spec.EngineVersion != nil {
+		res.SetEngineVersion(*desired.ko.Spec.EngineVersion)
+	}
+
+	// Determine if we are trying to scale up an instance
+	scaleUpUpdate := false
+	if delta.DifferentAt("Spec.NodeType") && desired.ko.Spec.NodeType != nil {
+		for _, instance := range desired.ko.Status.AllowedScaleUpNodeTypes {
+			if *instance == *desired.ko.Spec.NodeType {
+				scaleUpUpdate = true
+				break
+			}
+		}
+	}
```

Comment:
This little complexity in the API kind of shows why only adding deltas to the Update types is non-trivial. We can only set some fields conditional on others. We can't exactly add a delta condition to every field, and copy the field contents over 1:1.

Reply:
I agree, we it is not trivial to generate the update payload logic lot of complexity involved.

Reply:
I agree, we it is not trivial to generate the update payload logic lot of complexity involved.